File: backend/src/room/room.rs
use serde::Serialize;

use crate::game::types::ManSpriteKey;
use crate::state::RoomListItem;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientDataForRoomClient {
    pub client_id: u32,
    pub client_name: String,
    pub is_ready: bool,
    pub is_host: bool,
    pub man_sprite_key: ManSpriteKey,
}

pub struct ClientDataForState {
    pub id: u32,
    pub name: String,
    pub is_ready: bool,
    pub is_host: bool,
    pub man_sprite_key: ManSpriteKey,
}

pub struct RoomState {
    pub id: u32,
    pub players: Vec<ClientDataForState>,
    pub opened_at: std::time::Instant,
    pub map_id: u32,
}

impl RoomState {
    pub fn new(id: u32, host_id: u32, host_name: String) -> Self {
        Self {
            id,
            players: vec![ClientDataForState {
                id: host_id,
                name: host_name,
                is_ready: false,
                is_host: true,
                man_sprite_key: ManSpriteKey::Man1,
            }],
            opened_at: std::time::Instant::now(),
            map_id: 1,
        }
    }

    pub fn to_list_item(&self) -> RoomListItem {
        RoomListItem {
            id: self.id,
            current_player_num: self.players.len(),
            opened_second: self.opened_at.elapsed().as_secs(),
            map_id: self.map_id,
        }
    }

    /// Snapshot of players for sending to clients.
    pub fn to_client_data(&self) -> Vec<ClientDataForRoomClient> {
        self.players
            .iter()
            .map(|p| ClientDataForRoomClient {
                client_id: p.id,
                client_name: p.name.clone(),
                is_ready: p.is_ready,
                is_host: p.is_host,
                man_sprite_key: p.man_sprite_key.clone(),
            })
            .collect()
    }
}

File: backend2/room/room.go
package room

import (
	"time"

	"bomberman/backend2/types"
)

type RoomState struct {
	ID       uint32
	Players  []types.PlayerInfo
	OpenedAt time.Time
	MapID    uint32
}

func NewRoomState(id uint32, hostID uint32, hostName string) *RoomState {
	return &RoomState{
		ID: id,
		Players: []types.PlayerInfo{
			{
				ID:           hostID,
				Name:         hostName,
				IsReady:      false,
				IsHost:       true,
				ManSpriteKey: types.Man1,
			},
		},
		OpenedAt: time.Now(),
		MapID:    1,
	}
}

func (r *RoomState) ToListItem() types.RoomListItem {
	return types.RoomListItem{
		ID:               r.ID,
		CurrentPlayerNum: len(r.Players),
		OpenedSecond:     uint64(time.Since(r.OpenedAt).Seconds()),
		MapID:            r.MapID,
	}
}

func (r *RoomState) ToClientData() []types.ClientDataForRoomClient {
	result := make([]types.ClientDataForRoomClient, len(r.Players))
	for i, p := range r.Players {
		result[i] = types.ClientDataForRoomClient{
			ClientID:     p.ID,
			ClientName:   p.Name,
			IsReady:      p.IsReady,
			IsHost:       p.IsHost,
			ManSpriteKey: p.ManSpriteKey,
		}
	}
	return result
}
